enova/serving/backend/injector.py

```diff
@@ -0,0 +1,107 @@
+        if kwargs.get("headers") is not None:
+            headers = kwargs.get("headers")
+
+        data = None
+        if kwargs.get("data") is not None:
+            data = kwargs.get("data")
+
+        if api_key:
+            headers["Authorization"] = f"Bearer {api_key}"
+        
+
+        start_time = kwargs.get("start_time")
+        sleep_duration = 0
+        if start_time is not None:
+            start_timestamp = datetime2timestamp(start_time)
+            now_timestamp = time.time()
+            if start_timestamp < now_timestamp:
+                start_timestamp = now_timestamp
+                LOGGER.info("[Traffic Injector] start time before current time, use current time as start time")
+
+            sleep_duration = start_timestamp - now_timestamp
+            end_time = kwargs.get("end_time")
+            if end_time is not None:
+                end_timestamp = datetime2timestamp(end_time)
+                if start_timestamp >= end_timestamp:
+                    raise TestStartError("[Traffic Injector] invalid end time: end time before start time")
+
+                duration = end_timestamp - start_timestamp
+
+        if method.lower() == "get":
+            request_method = requests.get
+        elif method.lower() == "post":
+            request_method = requests.post
+        else:
+            raise TestStartError("[Traffic Injector] invalid method: method not supported")
+
+        dataset_name = str.lower(data)
+        dataset_path = f"{get_enova_path()}/template/deployment/docker-compose/traffic-injector/data/{dataset_name}.csv"
+        with open(dataset_path) as dataset_in:
+            reader = csv.DictReader(dataset_in)
+            dataset = [s for s in reader]
+            for i, sample in enumerate(dataset):
+                # NOTE: only fits for OpenAI API server
+                if sample.get("answer"):
+                    del dataset[i]["answer"]
+                if sample.get("id"):
+                    del dataset[i]["id"]
+                if not sample.get("prompt") and sample.get("question"):
+                    dataset[i]["prompt"] = dataset[i].pop("question", None)
+                dataset[i]["model"] = model
+                dataset[i]["max_tokens"] = max_tokens
+                dataset[i]["temperature"] = temperature
+        dataset_iter = itertools.cycle(dataset) # NOTE: match JMeter's recycle:True
+
+        # initialize timer
+        if timer['type'] == Distribution.POISSON.value:
+            delay_iter = iter(np.random.poisson(timer["lambda"],duration))
+        elif timer['type'] == Distribution.NORMAL.value:
+            delay_iter = iter(np.random.normal(timer["mean"], timer["std"], duration))
+        else:
+            raise TestStartError(f"VanillaTrafficInjector: argument 'timer' provided is invalid, got {timer} instead")
+        def make_request(request_func, url, headers, json_data):
+            request_func(url, headers=headers, json=json_data)
```

Comment:
增加 try exception 捕捉错误并打印哈。 不然出错都不知道为什么啦

Reply:
这方面我计划是在调用injector这层捕捉（ https://github.com/Emerging-AI/ENOVA/blob/9b6dcac6157a099494834952033aa6931bcc9fc2/enova/serving/backend/base.py#L87C1-L103C29 ）： 
```suggestion
            try:
                injector = VanillaTrafficInjector()
                injector.run(
                    host=host,
                    port=port,
                    path=path,
                    method="post",
                    duration=int(pd.Timedelta(str(params.duration)+params.duration_unit).total_seconds()),
                    timer=timer,
                    data=params.data_set,
                    headers={"Content-Type":"application/json"},
                    model=self.model,
                )
                return True
            except TestStartError as e:
                LOGGER.error(e.message)
                return False
```